### PR TITLE
add new API to flush the internal page cache

### PR DESCRIPTION
**Pull request by Wenzel**
A few weeks ago, i opened an issue on libvmi : #464 

The solution was to flush the internal page cache, which is not accessible from any exposed API.

This PR exposes an API to flush the cache when the user thinks the memory layout has changed.

It worked for me, but maybe it will need more polishing from your side.
give me your feedback :)

### CODE REVIEW

**Comment by awsaba** [comment]
this is ok to test

**Comment by smaresca** [comment]
@Wenzel the failed due to:
> ../libvmi/check_libvmi-cache.o: In function `page_cache_flush':
> /home/jenkins/workspace/PR_LibVMI_KVM_Test_Winxpsp2/tests/../libvmi/cache.c:670: undefined reference to `memory_cache_flush'
> /home/jenkins/workspace/PR_LibVMI_KVM_Test_Winxpsp2/tests/../libvmi/cache.c:670: undefined reference to `memory_cache_flush'
> collect2: error: ld returned 1 exit status
> Makefile:590: recipe for target 'check_libvmi' failed

File: libvmi/cache.c [hunk 1]
```diff
@@ -663,6 +664,13 @@ v2p_cache_flush(
     dbprint(VMI_DEBUG_V2PCACHE, "--V2P cache flushed\n");
 }
 
+void
+page_cache_flush(
```
Comment (tklengyel) [inline libvmi/cache.c]: What's the point of having this extra function?

File: libvmi/cache.c [hunk 2]
```diff
@@ -733,3 +741,10 @@ vmi_v2pcache_flush(
 {
     return v2p_cache_flush(vmi, dtb);
 }
+
+void
+vmi_pagecache_flush(
+    vmi_instance_t vmi)
+{
+    return page_cache_flush(vmi);
```
Comment (tklengyel) [inline libvmi/cache.c]: You could just as well call `memory_cache_flush` here.

**Comment by tklengyel** [comment]
Evidently the KVM driver is lacking in comparison to Xen as what happens is that LibVMI reads the whole page into an internal buffer and caches it. Obviously this buffer will go out-of-sync sooner or later if the VM is running. So unless the VM is completely paused for the entire duration of the LibVMI instance, we can't guarantee that the memory stored in the cache is correct.

So this API addition doesn't really solve the problem and will not be evident that the user has to do the following:
1. pause vm
2. flush page cache
3. read/write
4. resume vm

If you don't pause the entire VM then the memory could go out-of-sync at any point. Furthermore, this is obviously a brute way of ensuring sanity as we don't know if we really need to flush the page or not (it might not have changed after all). But since the cache would have to be flushed every time you pause, then having the cache itself doesn't really get you much. You could just as well disable it and read the entire page in for yourself in your application when you pause the VM.

A much much better way would be to do away with reading in and caching the entire page to begin with. If we could get the KVM driver to provide us with a handle to the page like Xen does then we wouldn't have to worry about the buffer going out-of-sync and accessing it would also be faster (ie. zero-copy).

Bottom line, the current KVM memory access patches were not designed with running VMs in mind and this patch only provides a workaround that will be far from obvious and/or optimal.

**Comment by Wenzel** [comment]
This is how i'm currently using Libvmi with KVM:
1. run the vm
2. wait until the next event
3. event received, vm is paused
4. read/write the memory, decode data
5. continue the vm execution

I could have disabled the internal page cache with `./configure --disable-page-cache`, and it works ofc.

> But since the cache would have to be flushed every time you pause, then having the cache itself doesn't really get you much.

Since i'm doing a lot of read/write operations on the same page while processing the event, i need it to be cached by Libvmi to keep a fast memory access.

> You could just as well disable it and read the entire page in for yourself in your application when you pause the VM.

I prefer to keep these low level operations on the Libvmi level.

> If we could get the KVM driver to provide us with a handle to the page like Xen does then we wouldn't have to worry about the buffer going out-of-sync and accessing it would also be faster (ie. zero-copy).

What do you mean by Xen providing you a `handle` to the page ?
What operations can you do with it ?

Also, what about using EPT to monitor the pages that have been written during the vm execution ?
> When accessed and dirty flags for EPT are enabled, software can track writes to guest-physical addresses using a feature called `page-modification logging`

It could be a solution for the new design of the Libvmi KVM driver.

**Comment by tklengyel** [comment]
@Wenzel: the handle the Xen driver gives you is akin to a file handler. Effectively Xen maps the memory page from the target VM into the memory space of the process that uses LibVMI. That way the same physical memory is visible in both without having to pause and copy the page to a buffer. This would be the ideal setup for KVM as well.

Doing tricks with EPT is possible, ie. notify LibVMI when one of the pages in the cache have gone dirty, but IMHO that adds complexity. It can also leave the VM in an unstable state in case the process crashes and LibVMI's cleanup isn't called. It also adds a lot of overhead of trapping out every time the page changes when we might not even need the page in the cache anymore.

**Comment by awsaba** [comment]
I missed the part of the original issue that this was KVM specific, so now I'm kind of torn about whether this should make it into the external API.

RE: EPT tricks, the reread of the page could be limited to only happening on the first read/write after the violation, but as @tklengyel states, that is a lot of complexity that would probably not make it into libvmi when the alternatives seem to be:

- Fix up the QEMU memaccess patch
- Use Xen

**Comment by noxdafox** [comment]
@Wenzel maybe worth creating an issue on [KVM-VMI/kvm](https://github.com/KVM-VMI/kvm) describing the memory mapping feature @tklengyel is mentioning?

**Comment by awsaba** [comment]
@noxdafox I don't have links to the discussions handy, but attempts have been made to mainline something like [this patch](tools/qemu-kvm-patch/kvm-qemu-2.4.0.1-libvmi.patch) have been made before and haven't made it.

**Comment by Wenzel** [comment]
Designing a new approach to allow KVM to give a handle on a memory page like Xen does would be the way to go here.
However, this could easily take months or more, depending on how KVM manipulates the memory under the hood, and how to get an access to it without crashing the guest. 

In the meantime, would it be possible to still have the memory cache flushed ?

Like you said above, there is no need to add a new API, since we can flush when the user wants to flush the `v2p` cache:

~~~C
void
vmi_v2pcache_flush(
    vmi_instance_t vmi,
    addr_t dtb)
{
    memory_cache_flush(vmi);
    return v2p_cache_flush(vmi, dtb);
}
~~~

What does it changes on the Xen side ?
How do you know that the internal page cache needs to be flushed for Xen ?

**Comment by tklengyel** [comment]
On Xen the internal page-cache rarely needs to get flushed. The only consideration there is how many handles do you want to keep open at any given time as they don't go stale. So it's only a memory use issue as you don't want to have hundreds of page handles active all the time.

Flushing it as part of v2p flush doesn't really make much sense, I would prefer to have a standalone API for this. As I said, I'm not totally against this API addition, it is just a half-measure that doesn't really solve the problem.

**Comment by tklengyel** [comment]
So a quick read of https://lwn.net/Articles/658511 tells me what we should do is have QEMU share its `vmfd` with LibVMI and then have the KVM kernel module support an ioctl for `KVM_GET_USER_MEMORY_REGION`, similarly to the existing `KVM_SET_USER_MEMORY_REGION`. That would allow us to get the memory mapped into the LibVMI process without having to resort to actually copying the memory via qmp.

**Comment by tklengyel** [comment]
I think it is safe to close this PR for now, @Wenzel?
